testres/testres.1
.\"	$Id$
.\"
.\" Copyright (c) 2018 Sergey Bronnikov
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: November 5 2018 $
.Dt TESTRES 1
.Os
.Sh NAME
.Nm testres
.Nd A hyperfast web frontend for software testing results written in C.
.Sh SYNOPSIS
.Nm
.Op Fl h
.Op Fl s Ar file
.Op Fl s Ar directory
.Op Fl v
.Sh DESCRIPTION
The
.Nm
utility knits together software testing reports.
It helps to store and analyze them.
.Nm
is able to read software testing reports in three formats: JUnit, SubUnit
(versions 1 and 2) and Test Anything Protocol.
It supports two modes: console and Web application.
.Sh USAGE
.Bl -tag -width Ds
.It Pa Console application
shows a summary of multiple reports in a directory and detailed summary of a
single report.
.It Pa Web application
works as a CGI application.
Shows index of multiple reports, separate reports and supports searching.
.El
.Sh FORMATS
.Nm
supports three the most popular software testing report formats: TAP
(TestAnythingProtocol), SubUnit (version 1 and 2) and JUnit.
.Bl -tag -width Ds
.It Pa JUnit "   "
The JUnit XML Report output comes from a build tool called Nant, as opposed to
the JUnit project itself - thus it can be a little tricky to nail down an
official spec for the format, even though it's widely adopted and used.
There have been a number of attempts to codify the schema, first off there is
an XSD for JUnit.
.It Pa Test Anything Protocol
is a simple text-based interface between testing modules a test harness.
.It Pa SubUnit
is a streaming protocol for test results.
It allows communication between unit tests and a test harness.
Originally developed for unit testing in 2005 by Robert Collins.
There are two major revisions of the protocol.
Version 1 was trivially human readable but had significant defects.
Version 1.1 added encapsulation of binary streams which mitigated some of the
issues but the core remained.
Version 2 shares many of the good characteristics of Version 1 and it still
models xUnit style test execution.
.El
.Sh FILES
.Bl -tag -width "/var/www/conf/testres.css" -compact
.It Pa /var/www/conf/testres.css
CSS style sheet.
.It Pa /var/www/conf/favicon.ico
favicon.
.El
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
.Bd -literal

TEST REPORT (SUBUNIT_V1)
ID: b09b24ebda9e4725552bf89f9afdd9d04ab1cab
CREATED ON: Thu Jan  1 03:00:00 1970
FILE: web/subunit_v1.subunit

SUITE: noname
        PASS test_add.TestAdd.test_add_control_dir(pre-views)
        PASS test_add.TestAdd.test_add_control_dir(view-aware)
        PASS test_add.TestAdd.test_add_dry_run(pre-views)
        PASS test_add.TestAdd.test_add_dry_run(view-aware)

.Ed
.Sh STANDARDS
.Rs
.%A M. G. Schwern, A. Lester, A. Armstrong
.%D 2003-2007
.%T TAP 13 - The Test Anything Protocol v13
.Re
.Pp
.Rs
.%A R. Collins
.%D August 2009
.%T SubUnit: A streaming protocol for test results
.Re
.Pp
.Rs
.%A D. Robinson, K. Coar
.%D October 2004
.%R RFC 3875
.%T The Common Gateway Interface (CGI) Version 1.1
.Re
.Sh AUTHORS
.An -nosplit
The
.Nm
utility was written by
.An Sergey Bronnikov
.Sh CAVEATS
Supported formats uses different names and types of test statuses.
For simplification
.Nm
number of statuses limited by PASS, FAIL and SKIP in report summary. These
statuses aggregates other appropriate types of statuses in supported formats.

TODO: Bail out processing
TODO: APFD (Average Percentage of Faults Detected)
A higher APFD value indicates faster (or  better) fault detection rate.
TODO: TAP report files must have .tap extension, SubUnit - .subunit extension
and JUnit must have .xml extension.


testres/testres.c
/*
 * Copyright © 2018 Sergey Bronnikov
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  1. Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *
 *  2. Redistributions in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *     and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <err.h>
#include <fcntl.h>
#include <libgen.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>

#include "parse_common.h"
#include "ui_console.h"
#include "ui_http.h"

char version[1024];

void 
usage(char *path)
{
	char *name = basename(path);
	fprintf(stderr, "Usage: %s [-s file | directory] [-h|-v]\n", name);
}

int 
main(int argc, char *argv[])
{
	char *path = (char *) NULL;
	const char *stylesheet = "/testres.css";
	int opt = 0;

#ifdef __OpenBSD__
	if (pledge("stdio rpath", NULL) == -1) {
	    warn("pledge");
	    return EXIT_FAILURE;
	}
#endif /* __OpenBSD__ */

	snprintf(version, sizeof(version), "%s %s", __DATE__, __TIME__);
	while ((opt = getopt(argc, argv, "vhs:")) != -1) {
		switch (opt) {
		case 'h':
			usage(argv[0]);
			return 0;
		case 'v':
			printf("Build-date: %s\n", version);
			return 0;
		case 's':
			path = optarg;
			break;
		default:	/* '?' */
			usage(argv[0]);
			return 1;
		}
	}

	if (argc == 1) {
		usage(argv[0]);
		return 1;
	}

    	if (path == (char*)NULL) {
		perror("specified path is empty");
		return 1;
	}

	struct stat path_st;
	if (stat(path, &path_st) == -1) {
	   perror("cannot open specified path");
	   return 1;
	}

	char *query_string = getenv("QUERY_STRING");
	tailq_report *report_item;
	if (S_ISREG(path_st.st_mode)) {
	   report_item = process_file(path);
	   if (query_string != NULL) {
	      print_html_headers(version, stylesheet);
	      print_html_report(report_item);
	      print_html_footer(version);
	   } else {
	      /* print_report_prof(report_item); */
	      print_report(report_item);
 	   }
	   free_report(report_item);
	   return 0;
	}

	struct reportq *reports;
	reports = process_dir(path);
	if (query_string == NULL) {
	   print_reports(reports);
	   free_reports(reports);
	   return 0;
 	}

	if (strcmp(query_string, "index") == 0) {
	   print_html_headers(version, stylesheet);
	   print_html_reports(reports);
	   print_html_footer(version);
	   free_reports(reports);
	   return 0;
	}

    	if (strcmp(strtok(query_string, "="), "show") == 0) {
	   const char *report_id = strtok(NULL, "=");
	   struct tailq_report *report;
	   if ((report = is_report_exists(reports, report_id)) != NULL) {
	      print_html_headers(version, stylesheet);
	      print_html_report(report);
	      print_html_footer(version);
	      free_report(report);
	      free_reports(reports);
	   } else {
	      print_html_headers(version, stylesheet);
	      printf("not found\n");
	      print_html_footer(version);
	   }
	}
	return 0;
}
